File: labs-algo/1/F.java
import java.util.ArrayList;
import java.util.Scanner;

/**
 * @author Saveliy Bakturin
 *
 * Don't write off, if you don't wanna be banned!
 */

public class F {
    public static void insert(final ArrayList<Long> heap, final long x) {
        heap.add(x);
        int i = heap.size() - 1;
        while (heap.get(i) > heap.get((i - 1) / 2)) {
            long t = heap.get(i);
            heap.set(i, heap.get((i - 1) / 2));
            heap.set((i - 1) / 2, t);
            i = (i - 1) / 2;
        }
    }

    public static long extract(final ArrayList<Long> heap) {
        final long extracted = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        int index = 0;
        while (2 * index + 1 < heap.size()) {
            int left = 2 * index + 1;
            int right = 2 * index + 2;
            int j = left;
            if (right < heap.size() && heap.get(right) > heap.get(left)) {
                j = right;
            }
            if (heap.get(index) >= heap.get(j)) {
                break;
            }
            long t = heap.get(index);
            heap.set(index, heap.get(j));
            heap.set(j, t);
            index = j;
        }
        heap.remove(heap.size() - 1);
        return extracted;
    }

    public static void main(final String... args) {
        final Scanner in = new Scanner(System.in);
        final int n = in.nextInt();
        in.nextLine();
        final ArrayList<Long> heap = new ArrayList<>();
        final ArrayList<Long> answer = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            final String[] s = in.nextLine().split(" ");
            switch (s[0]) {
                case "0":
                    insert(heap, Long.parseLong(s[1]));
                    break;
                case "1":
                    answer.add(extract(heap));
                    break;
            }
        }
        in.close();
        for (final Long integer : answer) {
            System.out.println(integer);
        }
    }
}

File: labs/algo/1/F.cpp
#include <cstdio>
#include <vector>

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

class Heap {
private:
	std::vector<int> v;

public:
	Heap() = default;
	~Heap() = default;

	void insert(int const x);
	int extract();
};

void Heap::insert(int const x) {
	v.push_back(x);
	int i = v.size() - 1;
	while (v[i] > v[(i - 1) / 2]) {
		std::swap(v[i], v[(i - 1) / 2]);
		i = (i - 1) / 2;
	}
}

int Heap::extract() {
	int const max = v[0];
	v[0] = v[v.size() - 1];
	int i = 0;
	while (2 * i + 1 < v.size()) {
		int left = 2 * i + 1;
		int right = 2 * i + 2;
		int j = left;
		if (right < v.size() && v[right] > v[left]) {
			j = right;
		}
		if (v[i] >= v[j]) {
			break;
		}
		std::swap(v[i], v[j]);
		i = j;
	}
	v.resize(v.size() - 1);
	return max;
}

int main() {
	int n;
	scanf("%i", &n);
	Heap heap;
	for (int i = 0; i < n; i++) {
		int op;
		scanf("%i", &op);
		if (op == 0) {
			int val;
			scanf("%i", &val);
			heap.insert(val);
		} else {
			printf("%i\n", heap.extract());
		}
	}
}
